From a2be564f252fed2cf63f1bb2f0aed2d6e6034791 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 05 May 2022 08:23:07 +0300
Subject: [PATCH] [SW-74093] gaudi2: skip snps wa based on efuse

Based on efuse value of EFUSE_SYSCFG_WA_SNPS_CASE_01038891,
ppboot needs to skip configuration of synopsys work around
during pcie link bring up.

Change-Id: I18b05b179f12ce8e22b61a6d929dfedcaa705bc4
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 6d7bd40..df3df71 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -498,6 +498,16 @@
 #define EFUSE_SYSCFG_WA_SNPS_CASE_01038891    BIT(23)
 #define EFUSE_SYSCFG_PCIE_GEN_LIMIT           BIT(18)
 
+bool pcie_skip_wa_snps(void)
+{
+	uint32_t config_fuse = sp_sram_readl(EFUSE_SYSCFG_OFFSET);
+
+	if ((config_fuse & EFUSE_SYSCFG_WA_SNPS_CASE_01038891) == 0)
+		return false;
+
+	return true;
+}
+
 static void pcie_wa_snps_case_01038891(uint32_t boot_strap)
 {
 	uint8_t separate_reference = 0;
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 93ea7c5..a5ece51 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -444,6 +444,11 @@
 	return !!(boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H_SRIS_MODE_MASK);
 }
 
+__weak bool pcie_skip_wa_snps(void)
+{
+	return false;
+}
+
 void pcie_wa_snps_case_01038891(void)
 {
 	uint8_t separate_reference = 0;
@@ -464,16 +469,18 @@
 	/* spread spectum clock enable */
 	ssc_en = get_ssc_en_from_strap(boot_strap);
 
-	/* TODO: Check fuse to see if SNPS WA needs to be enabled  */
+	if (pcie_skip_wa_snps())
+		return;
+
 	if (ssc_en) {
 		pcie_phy_set_mpll_a_b_ssc_enable();
 
 		regmap_writel(0x5000, mmPCIE_PHY_EXT_MPLLB_SSC_PEAK);
 		regmap_writel(0x19CE,
-				mmPCIE_PHY_EXT_MPLLB_SSC_STEP_SIZE);
+			      mmPCIE_PHY_EXT_MPLLB_SSC_STEP_SIZE);
 		regmap_writel(0x6666, mmPCIE_PHY_EXT_MPLLA_SSC_PEAK);
 		regmap_writel(0x2108,
-				mmPCIE_PHY_EXT_MPLLA_SSC_STEP_SIZE);
+			      mmPCIE_PHY_EXT_MPLLA_SSC_STEP_SIZE);
 	}
 
 	pcie_phy_set_ext_tx_rx_ctrl();
